ee/clickhouse/migrations/0021_session_recording_events_materialize_full_snapshot.py

```diff
@@ -0,0 +1,50 @@
+from infi.clickhouse_orm import migrations
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.materialized_columns.columns import get_materialized_columns
+from ee.clickhouse.sql.session_recording_events import SESSION_RECORDING_EVENTS_TABLE
+from posthog.settings import CLICKHOUSE_CLUSTER, CLICKHOUSE_REPLICATION
+
+
+def create_has_full_snapshot_materialized_column(database):
+
+    if "has_full_snapshot" in get_materialized_columns(SESSION_RECORDING_EVENTS_TABLE, use_cache=False):
+        # Field is already materialized
+        return
+
+    if CLICKHOUSE_REPLICATION:
+        sync_execute(
+            f"""
+            ALTER TABLE sharded_{SESSION_RECORDING_EVENTS_TABLE}
+            ON CLUSTER {CLICKHOUSE_CLUSTER}
+            ADD COLUMN IF NOT EXISTS
+            has_full_snapshot BOOLEAN MATERIALIZED JSONExtractBool(snapshot_data, 'has_full_snapshot')
+        """
+        )
+        sync_execute(
+            f"""
+            ALTER TABLE {SESSION_RECORDING_EVENTS_TABLE}
+            ON CLUSTER {CLICKHOUSE_CLUSTER}
+            ADD COLUMN IF NOT EXISTS
+            has_full_snapshot BOOLEAN
```

Comment:
I'm following patterns here, but I don't really understand why we only materialize the column on the sharded table.

Reply:
This is to adapt to the cloud clickhouse schema. 

Pretty sure this will blow up on cloud. Check out `system.tables` table on cloud/metabase.

Reply:
I started with the approach of just calling `materialize`, but this case required updating that function and `backfill_materialized_columns` to support:
* Extracting from `snapshot_data` instead of `properties`
* Having a column type `BOOLEAN` instead of `VARCHAR`
* Extracting using `JSONExtractBool` instead of `JSONExtractRaw`

After making the changes to `materialize` to support changing these things, it felt like it was adding a bunch of "cruft" to handle this 1 case, and given the importance of this function, it felt wrong to over complicate it.

I could be convinced to go back to that approach, but that was my rational.

Reply:
> Pretty sure this will blow up on cloud. Check out system.tables table on cloud/metabase. It also doesn't behave the same as the materialize function.

Spent some time on this. From what I see, I think this is correct. It looks like `session_recording_events` and `events` are both setup the same way with `sharded_` versions according to `system.tables`. And I think this migration mimics how the `materialize` function works for `events`. Are you seeing something specific that makes you think this will blow up?